File: developerRobot/src/main/native/cpp/Robot.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <frc/Encoder.h>
#include <frc/TimedRobot.h>
#include <frc/controller/PIDController.h>
#include <frc/controller/SimpleMotorFeedforward.h>
#include <frc/motorcontrol/PWMSparkMax.h>
#include <frc/trajectory/ExponentialProfile.h>

#include <units/length.h>
#include <units/velocity.h>
#include <units/voltage.h>

class Robot : public frc::TimedRobot {
 public:
  inline static constexpr units::second_t dt = 10_ms;
  inline static constexpr units::second_t offset = 5_ms;

  inline static constexpr auto Ks = 0.1_V;
  inline static constexpr auto Kv = 3_V / 1_mps;
  inline static constexpr auto Ka = 0.5_V / 1_mps_sq;

  inline static constexpr auto Kp = 0.1_V / 1_m;
  inline static constexpr auto Kd = 0.1_V / 1_mps;

  frc::Encoder encoder{0, 1};
  frc::PWMSparkMax motor{0};

  frc::ExponentialProfile<units::meter, units::volt> profile{{10_V, Kv, Ka}};

  frc::SimpleMotorFeedforward<units::meter> feedforward{Ks, Kv, Ka};
  frc::PIDController feedback{Kp.value(), 0.0, Kd.value(), dt};

  decltype(profile)::State currentSetpoint;
  decltype(profile)::State goal;

  Robot() {
    AddPeriodic([this] { Controller(); }, dt, offset);
  }

  void SetGoal(units::meter_t goal) { this->goal.position = goal; }

  void Controller() {
    if (IsDisabled()) {
      return;
    }

    auto nextSetpoint = profile.Calculate(dt, currentSetpoint, goal);
    motor.SetVoltage(
        feedforward.Calculate(currentSetpoint.velocity, nextSetpoint.velocity,
                              dt) +
        units::volt_t{feedback.Calculate(encoder.GetDistance(),
                                         currentSetpoint.position.value())});
    currentSetpoint = nextSetpoint;
  }
};

int main() {
  return frc::StartRobot<Robot>();
}

File: developerRobot/src/main/java/frc/robot/Robot.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot;

import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.controller.SimpleMotorFeedforward;
import edu.wpi.first.math.trajectory.ExponentialProfile;
import edu.wpi.first.wpilibj.Encoder;
import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.motorcontrol.PWMSparkMax;

@SuppressWarnings({"MemberName", "MissingJavadocMethod"})
public class Robot extends TimedRobot {
  public static final double dt = 0.01;
  public static final double offset = 0.005;

  public static final double Ks = 0.1;
  public static final double Kv = 3.0;
  public static final double Ka = 0.5;

  public static final double Kp = 0.1;
  public static final double Kd = 0.1;

  public Encoder encoder = new Encoder(0, 1);
  public PWMSparkMax motor = new PWMSparkMax(0);

  public ExponentialProfile profile =
      new ExponentialProfile(ExponentialProfile.Constraints.fromCharacteristics(10.0, Kv, Ka));

  public SimpleMotorFeedforward feedforward = new SimpleMotorFeedforward(Ks, Kv, Ka);
  public PIDController feedback = new PIDController(Kp, 0.0, Kd, dt);

  public ExponentialProfile.State currentSetpoint = new ExponentialProfile.State();
  public ExponentialProfile.State goal = new ExponentialProfile.State();

  public Robot() {
    addPeriodic(this::controller, dt, offset);
  }

  public void setGoal(double goal) {
    this.goal.position = goal;
  }

  public void controller() {
    if (isDisabled()) {
      return;
    }

    var nextSetpoint = profile.calculate(dt, currentSetpoint, goal);
    motor.setVoltage(
        feedforward.calculate(currentSetpoint.velocity, nextSetpoint.velocity, dt)
            + feedback.calculate(encoder.getDistance(), currentSetpoint.position));
    currentSetpoint = nextSetpoint;
  }
}
